### PR TITLE
Feat product filters

### PR DESCRIPTION
**Pull request by ankit9015**
This PR has following features:
1. Product Filter including category, ratings, price, 
2. Product sort: on the basis of price
3. DisplaySection.jsx is controlling the products display grid
4. SideBar.jsx has all the filters
5. product-filter.jsx contains all the different filter functions and a compositeFilter that reduces all the filters

### CODE REVIEW

File: src/pages/Products/DisplaySection.jsx [hunk 1]
```diff
@@ -1,18 +1,30 @@
...
-      <h2 class="H2 font-bold line-height-lg">
+      <h2 className="H2 font-bold line-height-lg">
         Showing All Products{" "}
-        <span class="text-md m-s">(Showing 20 Products)</span>
+        <span className="text-md m-s">{`(Showing ${products.length} products) `}</span>
       </h2>
 
-      <div class="product-card-container">
+      <div className="product-card-container">
         {products.map((product) => (
```
Comment (atharvabhanage02) [inline src/pages/Products/DisplaySection.jsx]: You can map the products by checking if there are products available 
{ products && products.map } as this makes your code more maintainable

File: src/pages/Products/SideBar.jsx [hunk 2]
```diff
@@ -1,90 +1,171 @@
...
+  const [categories, setCategories] = useState([]);
+  const { productState, productDispatch } = useProductContext();
+
+  useEffect(() => {
+    (async () => {
+      try {
+        const response = await axios.get("/api/categories");
+        const categoriesList = response.data.categories;
+        setCategories((prev) => [...prev, ...categoriesList]);
+        // setLoading(false);
```
Comment (atharvabhanage02) [inline src/pages/Products/SideBar.jsx:23]: can remove comments before merging

File: src/utils/product-filter.jsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
+export const compositeFilter =
+  (state, ...filterFunctions) =>
+  (products) => {
+    // console.log(products);
```
Comment (atharvabhanage02) [inline src/utils/product-filter.jsx]: remove all the console.log once the checking has been done

**Review by atharvabhanage02 — APPROVED** [review]
Few Suggestions , else everything looks fine and in place

File: src/pages/Products/SideBar.jsx [hunk 4]
```diff
@@ -1,90 +1,171 @@
...
+              checked={productState.sort.order === "HIGHEST-FIRST"}
+              onChange={() =>
+                productDispatch({
+                  type: "SORT-PRODUCTS",
+                  payload: { sortBy: "price", order: "HIGHEST-FIRST" },
+                })
+              }
+            />
+            Highest first
+          </label>
```
Comment (Namrata-J) [inline src/pages/Products/SideBar.jsx:168]: good work here, you have broken down it into small components.

File: src/utils/product-filter.jsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  // console.log("inside", products);
+  const temp =
+    state.price.min === 0 && state.price.max === 5000
+      ? products
+      : products.filter(
+          ({ discountedPrice }) =>
+            discountedPrice >= state.price.min &&
+            discountedPrice <= state.price.max
+        );
+  // console.log("temp", temp);
```
Comment (Namrata-J) [inline src/utils/product-filter.jsx]: remove all the commented console.log's, it's not a good to have commented codes's and console log's in production code as well.

**Review by Namrata-J — APPROVED** [review]

File: src/App.js [hunk 6]
```diff
@@ -1,13 +1,16 @@
...
     <div className="App">
-      <PageLayout>
-        <Router />
-      </PageLayout>
+      <ProductProvider>
+        <PageLayout>
+          <Router />
+        </PageLayout>
+      </ProductProvider>
       {/* <Header></Header>
```
Comment (shaksham08) [inline src/App.js]: Remove commented code

**Review by shaksham08 — APPROVED** [review]
Code looks well structured ..just remove the commented code

**Pull request merged**